src/sage/cpython/atexit.pyx

```diff
@@ -144,51 +144,99 @@ cdef class restore_atexit:
         _set_exithandlers(self._exithandlers)
 
 from cpython.ref cimport PyObject
+import sys
 
-# Implement "_atexit_callbacks()" for each supported python version
+# Implement a uniform interface for getting atexit callbacks
 cdef extern from *:
     """
+    #ifndef Py_BUILD_CORE
     #define Py_BUILD_CORE
+    #endif
     #undef _PyGC_FINALIZED
     #include "internal/pycore_interp.h"
     #include "internal/pycore_pystate.h"
-    #if PY_VERSION_HEX >= 0x030c0000
-    // struct atexit_callback was renamed in 3.12 to atexit_py_callback
-    #define atexit_callback atexit_py_callback
-    #endif
-    static atexit_callback ** _atexit_callbacks(PyObject *self) {
+    
+    // Always define this struct for Cython's use
+    typedef struct {
+        PyObject *func;
+        PyObject *args;
+        PyObject *kwargs;
+    } atexit_callback_struct;
+    
+    #if PY_VERSION_HEX >= 0x030e0000
+    // Python 3.14+: atexit uses a PyList
+    static PyObject* get_atexit_callbacks_list(PyObject *self) {
         PyInterpreterState *interp = _PyInterpreterState_GET();
         struct atexit_state state = interp->atexit;
         return state.callbacks;
     }
+    
+    // Dummy function for Python 3.14+ (never called)
+    static atexit_callback_struct** get_atexit_callbacks_array(PyObject *self) {
+        PyErr_SetString(PyExc_RuntimeError, "Python >= 3.14 has no atexit array");
+        return NULL;
+    }
+    #else
+    // Python < 3.14: atexit uses C array
+    static atexit_callback_struct** get_atexit_callbacks_array(PyObject *self) {
+        PyInterpreterState *interp = _PyInterpreterState_GET();
+        struct atexit_state state = interp->atexit;
+        // Cast from atexit_callback** to our struct type
+        return (atexit_callback_struct**)state.callbacks;
+    }
+    
+    // Dummy function for Python < 3.14 (never called)
+    static PyObject* get_atexit_callbacks_list(PyObject *self) {
+        PyErr_SetString(PyExc_RuntimeError, "Python < 3.14 has no atexit list");
+        return NULL;
+    }
+    #endif
     """
-    ctypedef struct atexit_callback:
+    # Declare both functions - they exist in all Python versions (one is dummy)
+    PyObject* get_atexit_callbacks_list(object module)
```

Comment:
```suggestion
    object get_atexit_callbacks_list(object module)
```
This avoids the cast later, and will automatically get checked for NULL when called.